### PR TITLE
Workflow signals, queries, and more

### PR DESCRIPTION
**Pull request by cretz**
## What was changed

* Workflow signals
* Workflow queries
* Workflow cancellation
* Workflow timers
* Workflow utils like is replaying, random, wait condition, and current date
* Expose activity, workflow, signal, and query definitions for advanced users
* Fixes from PR of #29

The rest of workflow impl is coming in future PR(s). This also includes SDK core upgrade with multiple new things that will change some of the logic herein.

### CODE REVIEW

File: src/Temporalio/Activities/ActivityDefinition.cs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        /// </summary>
+        /// <param name="name">Activity name.</param>
+        /// <param name="del">Activity delegate.</param>
+        /// <returns>Activity definition.</returns>
+        /// <seealso cref="FromDelegate" />
+        public static ActivityDefinition CreateWithoutAttribute(string name, Delegate del)
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentException("Name required for activity");
```
Comment (cgillum) [inline src/Temporalio/Activities/ActivityDefinition.cs:57]: Consider providing the parameter name argument in these `ArgumentException` constructors. For example:
```suggestion
                throw new ArgumentException("Name required for activity", nameof(name));
```

File: src/Temporalio/Worker/ActivityWorker.cs [hunk 2]
```diff
@@ -665,7 +673,7 @@ public override void Init(ActivityOutboundInterceptor outbound)
                 object? result;
                 try
                 {
-                    result = input.Delegate.DynamicInvoke(input.Parameters);
+                    result = input.Delegate.DynamicInvoke(input.Args);
                 }
                 catch (TargetInvocationException e)
                 {
```
Comment (cgillum) [inline src/Temporalio/Worker/ActivityWorker.cs:679]: Not part of this PR, but is it correct to do `throw e.InnerException!` or is it better to do `ExceptionDispatchInfo.Capture(e.InnerException).Throw();`. More info: https://stackoverflow.com/a/17091351/2069

File: src/Temporalio/Worker/TemporalWorkerOptions.cs [hunk 3]
```diff
@@ -46,6 +46,22 @@ public TemporalWorkerOptions(string taskQueue)
...
         /// </summary>
         public IList<Type> Workflows { get; set; } = new List<Type>();
+
+        /// <summary>
+        /// Gets or sets additional activity definitions.
+        /// </summary>
+        /// <remarks>
+        /// Unless manual definitions are required, users should use <see cref="Activities" />.
+        /// </remarks>
+        public IList<Activities.ActivityDefinition> AdditionalActivityDefinitions { get; set; } = new List<Activities.ActivityDefinition>();
```
Comment (cgillum) [inline src/Temporalio/Worker/TemporalWorkerOptions.cs:56]: nit: It seems confusing to have a distinction between `Activities` and `AdditionalActivityDefinitions`. I understand that you factored the API this way so that you can have a list-based mechanism for registering implicitly named activities (and workflows) and explicitly named activities. However, the benefit of this design is not clear to me and I worry that the average developer wouldn't understand the motivation. I suspect it would be more natural to have two overloaded `AddActivity()` methods (`AddActivity(Delegate)` and `AddActivity(string, Delegate)`) and then a single `IReadOnlyList<ActivityDefinition> Activities { get; }` property to enable enumerating the definitions.

File: README.md [hunk 4]
```diff
@@ -341,6 +342,21 @@ Notes about the above code:
     method name.
   * This attribute is not inherited and therefore must be explicitly set on any override.
 
+#### Workflow Task Scheduling
```
Comment (cgillum) [inline README.md:345]: One question I have is how will developers unit test their workflows. This might be something to consider covering in the README as well.

**Review by cgillum — COMMENTED** [review]
Just a few observations. Feel free to ignore them.

File: src/Temporalio/Activities/ActivityDefinition.cs (hunk 1 above)
Comment (cretz) [reply src/Temporalio/Activities/ActivityDefinition.cs:57]: :+1: Will fix

File: src/Temporalio/Worker/ActivityWorker.cs (hunk 2 above)
Comment (cretz) [reply src/Temporalio/Worker/ActivityWorker.cs:679]: Ah, missed this one (you'll see I use that `ExceptionDispatchInfo` approach in other places). Will fix.

File: src/Temporalio/Worker/TemporalWorkerOptions.cs (hunk 3 above)
Comment (cretz) [reply src/Temporalio/Worker/TemporalWorkerOptions.cs:56]: I appreciate the simplicity of worker options as `new() { Activities = { ActivityClass.ActivityMethod } }` for the vast majority of users. I also intentionally made this unwieldy to discourage its use. Are there other ways to solve my goals of 1) making delegates with attributes the obvious preferred approach, and 2) making delegates with attributes easy to set as properties? I guess I could make a custom collection. Would that be a better approach?

I was expecting, with one of my samples, to show how to set activities as fully qualified method names using https://learn.microsoft.com/en-us/dotnet/core/extensions/configuration so I was hoping to keep everything properties I could one-day serialize into from JSON (granted I know that I'd have to manually resolve method references from strings during deserialization).

File: README.md (hunk 4 above)
Comment (cretz) [reply README.md:345]: Oh there is a ton to cover in the README, I just wanted to make this scheduling note so I didn't forget it :-)

Basically the README will look like Python's: https://github.com/temporalio/sdk-python#readme. There is gonna be a ton of stuff there coming in the next PR or two.

Specifically for testing workflows, the approach at https://github.com/temporalio/sdk-python#testing will also apply here. You will use `Temporalio.Testing.WorkflowEnvironment` class. These are actually more like integration tests since they run real servers as external processes, but they are fast, isolated, and give full functionality. The time-skipping test server is actually Graal-compiled from Java at https://github.com/temporalio/sdk-java that we use in TypeScript and Python today. In the future, if desired, we may offer a separate NuGet package with that test server as a DLL instead of lazy downloaded and run as a separate process.

**Comment by cretz** [comment]
Ignore failing Windows CI. Like last PR, this is still a problem with `dotnet format`: https://github.com/dotnet/format/issues/1800. If it goes on long enough, I'll (temporarily) disable code style in build.

File: src/Temporalio/Worker/TemporalWorkerOptions.cs (hunk 3 above)
Comment (cgillum) [reply src/Temporalio/Worker/TemporalWorkerOptions.cs:56]: I see. Supporting a JSON representation of worker options is an interesting idea. I can't immediately think of a better approach given these goals.

File: tests/Temporalio.Tests/Worker/WorkflowCodecHelperTests.cs [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+            await WorkflowCodecHelper.DecodeAsync(codec, act);
+            if (!payload().Metadata.ContainsKey("decoded"))
+            {
+                Assert.Fail($"Payload at path {ctx.Path} not decoded");
+            }
+        });
+    }
+
+    // Creates payloads as needed, null context if already seen
+    private static async Task CreateAndVisitPayload(
```
Comment (bergundy) [inline tests/Temporalio.Tests/Worker/WorkflowCodecHelperTests.cs:81]: Nice!

Comment (cretz) [reply tests/Temporalio.Tests/Worker/WorkflowCodecHelperTests.cs:81]: I knew you'd like this one (you may be the only one that will read/grok it). As expected, it helped me catch multiple issues.

File: src/Temporalio/Worker/WorkflowInstance.cs [hunk 6]
```diff
@@ -20,20 +23,30 @@ namespace Temporalio.Worker
...
-    internal class WorkflowInstance : TaskScheduler, IWorkflowInstance
+    internal class WorkflowInstance : TaskScheduler, IWorkflowInstance, IWorkflowContext
     {
         private readonly TaskFactory taskFactory;
-        private readonly WorkflowInstanceDetails details;
+        private readonly WorkflowDefinition defn;
+        private readonly IPayloadConverter payloadConverter;
+        private readonly IFailureConverter failureConverter;
+        private readonly Lazy<WorkflowInboundInterceptor> inbound;
+        private readonly Lazy<WorkflowOutboundInterceptor> outbound;
```
Comment (bergundy) [inline src/Temporalio/Worker/WorkflowInstance.cs:33]: Do these need to be lazy?
You'll always have interceptors AFAICT.

File: src/Temporalio/Worker/WorkflowInstance.cs [hunk 7]
```diff
@@ -20,20 +23,30 @@ namespace Temporalio.Worker
...
+        // Lazily created if asked for by user
+        private readonly Lazy<NotifyOnSetDictionary<string, WorkflowQueryDefinition>> mutableQueries;
+        // Lazily created if asked for by user
+        private readonly Lazy<NotifyOnSetDictionary<string, WorkflowSignalDefinition>> mutableSignals;
```
Comment (bergundy) [inline src/Temporalio/Worker/WorkflowInstance.cs:37]: Is the lazyness worth it?
What does it buy us? Is it just a general good practice?

File: src/Temporalio/Worker/WorkflowInstance.cs (hunk 6 above)
Comment (cretz) [reply src/Temporalio/Worker/WorkflowInstance.cs:33]: I should probably document this better, but it is important for me to run all user code _inside_ our scheduler task. The constructor of this class is not run inside there. By properly waiting to instantiate the workflow class and properly waiting to build/init interceptors until we are inside the first job processed, I can ensure that `Workflow.Info` will work in workflow constructors and interceptors. If I eagerly instantiated either of those now you couldn't get workflow info.

(this was lazy before, but manually so as a nullable property and I was able to move to this model as I reworked some other things)

File: src/Temporalio/Worker/WorkflowInstance.cs (hunk 7 above)
Comment (cretz) [reply src/Temporalio/Worker/WorkflowInstance.cs:37]: Last PR you dinged me on unnecessarily storing data when not needed, so this is kinda the same thing. Only users that ever request the `Workflow.Signals` property (so almost never) would create this. I am probably overdoing it a bit and could just create this dictionary on every workflow always. But technically this does save a few cycles and a few bytes (again though, probably negligible).

File: src/Temporalio/Worker/WorkflowInstance.cs [hunk 8]
```diff
@@ -191,7 +243,9 @@ public WorkflowActivationCompletion Activate(WorkflowActivation act)
             {
                 completion = new() { RunId = act.RunId, Successful = new() };
                 currentActivationException = null;
-                // We need to sort jobs
+                IsReplaying = act.IsReplaying;
+                UtcNow = act.Timestamp.ToDateTime();
+                // We need to sort jobs by notify, then signal, then non-query, then query
                 var jobs = act.Jobs.OrderBy(job =>
```
Comment (bergundy) [inline src/Temporalio/Worker/WorkflowInstance.cs:249]: nit: don't need this if you upgrade core, can get rid of it in the next PR

Comment (cretz) [reply src/Temporalio/Worker/WorkflowInstance.cs:249]: Yeah, there's a few things I have to change on next core upgrade. I am delaying it a bit while in mid-development on other things.

File: src/Temporalio/Worker/WorkflowInstance.cs [hunk 9]
```diff
@@ -278,7 +332,10 @@ public WorkflowActivationCompletion Activate(WorkflowActivation act)
                         i++;
                     }
                 }
-                return completion;
+                // Unset the completion
```
Comment (bergundy) [inline src/Temporalio/Worker/WorkflowInstance.cs:335]: Obviously not for this PR but I noticed you don't do this in Python.

File: src/Temporalio/Worker/WorkflowInstance.cs [hunk 10]
```diff
@@ -350,7 +407,16 @@ private void RunOnce(bool checkConditions)
                     }
                 }
 
-                // TODO(cretz): Check conditions
+                // Collect all condition sources to mark complete and then complete them. This
+                // avoids modify-during-iterate issues.
+                if (checkConditions)
```
Comment (bergundy) [inline src/Temporalio/Worker/WorkflowInstance.cs:412]: This is better than what we do in Python where we check conditions only after each batch of jobs.

I'd have to test but with this approach it seems like we're less likely to reach `A-B-A` situations where the condition reaches `B` and is immediately flipped back to `A` (e.g. if set by different signal handlers).

File: src/Temporalio/Worker/WorkflowInstance.cs (hunk 9 above)
Comment (cretz) [reply src/Temporalio/Worker/WorkflowInstance.cs:335]: That's because in python I considered it much less likely people would add commands outside of workflow execution. It's much easier to screw up and do it in .NET where we can't force the scheduler on you as well. But I can add it there, sure.

File: src/Temporalio/Worker/WorkflowInstance.cs (hunk 10 above)
Comment (cretz) [reply src/Temporalio/Worker/WorkflowInstance.cs:412]: This checks after each batch kinda too. Note the nested loops.

File: src/Temporalio/Worker/WorkflowInstance.cs (hunk 6 above)
Comment (bergundy) [reply src/Temporalio/Worker/WorkflowInstance.cs:33]: Might want to test that interceptor is instantiated with the proper context to make sure that contract is held.

Comment (cretz) [reply src/Temporalio/Worker/WorkflowInstance.cs:33]: Yeah, when I get around to my full interceptor testing, it will include confirming workflow info is available in all places (and for workflow class constructor for that matter)

File: src/Temporalio/Worker/WorkflowInstance.cs (hunk 10 above)
Comment (bergundy) [reply src/Temporalio/Worker/WorkflowInstance.cs:412]: We trigger `RunOnce` after applying every job, not once per batch.

Comment (bergundy) [reply src/Temporalio/Worker/WorkflowInstance.cs:412]: Which is the right behavior AFAICT but I have not confirmed there's an issue with the way it's done in Python.
In any case it's an edge case.

File: src/Temporalio/Worker/WorkflowInstance.cs [hunk 11]
```diff
@@ -422,51 +488,229 @@ private void Apply(WorkflowActivationJob job)
...
+            // Queue it up so it can run in workflow environment
+            _ = QueueNewTaskAsync(() =>
             {
-                var resultObj = await Inbound.ExecuteWorkflowAsync(new(
+                var origCmdCount = completion?.Successful?.Commands?.Count;
+                try
+                {
+                    // Find definition or fail
+                    var queries = mutableQueries.IsValueCreated ? mutableQueries.Value : defn.Queries;
+                    if (!queries.TryGetValue(query.QueryType, out var queryDefn))
```
Comment (bergundy) [inline src/Temporalio/Worker/WorkflowInstance.cs:541]: I didn't know about this cool .NET feature that `out` variables are available in the parent scope.

File: src/Temporalio/Worker/WorkflowInstance.cs [hunk 12]
```diff
@@ -422,51 +488,229 @@ private void Apply(WorkflowActivationJob job)
...
+        {
+            if (name != defn.Name)
+            {
+                throw new ArgumentException($"Signal name {name} doesn't match definition name {defn.Name}");
+            }
+            // Send all buffered signals
+            if (bufferedSignals.TryGetValue(name, out var buffered))
+            {
+                bufferedSignals.Remove(name);
+                buffered.ForEach(ApplySignalWorkflow);
```
Comment (bergundy) [inline src/Temporalio/Worker/WorkflowInstance.cs:658]: Can you confirm that these signals get a chance to be processed before the workflow execute function is resumed so no signals are lost in case the workflow attempts to complete?

File: src/Temporalio/Worker/WorkflowInstance.cs [hunk 13]
```diff
@@ -491,13 +736,115 @@ public override void Init(WorkflowOutboundInterceptor outbound)
...
+                    throw new ArgumentException("Delay duration cannot be less than 0.");
+                }
+                else if (delay == TimeSpan.Zero)
+                {
+                    delay = TimeSpan.FromMilliseconds(1);
+                }
+                var source = new TaskCompletionSource<object?>();
+                // Only create the command if not infinite. We use seq 0 to represent uncreated.
+                uint seq = 0;
+                if (delay != Timeout.InfiniteTimeSpan)
```
Comment (bergundy) [inline src/Temporalio/Worker/WorkflowInstance.cs:820]: Hmm.. only thing I'm worried about not creating the timer is that changing the duration from non infinite to infinite breaks determinism same way as it is in some other SDKs for changing from 0 to non 0, which we want to avoid.

**Review by bergundy — APPROVED** [review]
Looks great!
I mostly focused on the WorkflowInstance implementation. Skimmed some of the other stuff and skipped the workflow worker tests.

File: src/Temporalio/Worker/WorkflowInstance.cs (hunk 10 above)
Comment (cretz) [reply src/Temporalio/Worker/WorkflowInstance.cs:412]: Ah, yes, that is correct I forgot that I don't group jobs I just order them. I agree that batch + run once is a bit more confusing.

File: src/Temporalio/Worker/WorkflowInstance.cs (hunk 11 above)
Comment (cretz) [reply src/Temporalio/Worker/WorkflowInstance.cs:541]: Yeah, I found that out by accident a couple months ago when trying to shadow one. It makes sense though of course since it makes this pattern useful.

File: src/Temporalio/Worker/WorkflowInstance.cs (hunk 12 above)
Comment (cretz) [reply src/Temporalio/Worker/WorkflowInstance.cs:658]: Not following the question exactly. `OnSignalDefinitionAdded` is for when a user adds a signal handler manually and we are draining these inline at that time. You are already running the workflow code when you're adding a signal handler manually, there is no separate resumption.

File: src/Temporalio/Worker/WorkflowInstance.cs (hunk 13 above)
Comment (cretz) [reply src/Temporalio/Worker/WorkflowInstance.cs:820]: This is called out in the docs of `Workflow.Delay`:

```csharp
        /// The <c>delay</c> value can be <see cref="Timeout.Infinite" /> or
        /// <see cref="Timeout.InfiniteTimeSpan" /> but otherwise cannot be negative. A server-side
        /// timer is not created for infinite delays, so it is non-deterministic to change a timer
        /// to/from infinite from/to an actual value.
```

I think this is a reasonable tradeoff because in .NET people use infinite timers to create tasks that are just resolved on cancellation. Cancellation tokens don't really have a "task" you can wait on and we will probably forbid threaded `WaitHandle` things, so `await Workflow.DelayAsync(Timeout.Infinite, Workflow.CancellationToken)` may be a common pattern. We can try to avoid with helpers of course.

I could make a 100 year (or max duration or whatever) timer on the server, sure, but is it worth it just to allow that rare case of allowing people to change the timer duration later?

File: src/Temporalio/Activities/ActivityDefinition.cs (hunk 1 above)
Comment (cretz) [reply src/Temporalio/Activities/ActivityDefinition.cs:57]: Fixed in next PR

**Comment by cretz** [comment]
I am merging this so I can prep the next PR at #34. Any conversations can continue here or be added there.

(to external reviewers reading, we chose the multi-chopped-up PR approach while doing this initial dev for easier review while not slowing down the actual work, but it means it's best for us to address concern in future PRs as they are stacked)

File: src/Temporalio/Worker/WorkflowInstance.cs (hunk 12 above)
Comment (bergundy) [reply src/Temporalio/Worker/WorkflowInstance.cs:658]: So any signal handler that I add dynamically is immediately invoked?
In this example will the workflow be prevented from completing?

(pseudocode)
```
bool preventFromCompleting = false;
Workflow.Signals.Add(BlaSignal, () =>
{
  preventFromCompleting = true;
});

if (preventFromCompleting)
{
  // do something
}
```

File: src/Temporalio/Worker/WorkflowInstance.cs (hunk 13 above)
Comment (bergundy) [reply src/Temporalio/Worker/WorkflowInstance.cs:820]: Okay, let's keep it, it's well documented.
Some people will trip on this but I agree it's a reasonable tradeoff.

File: src/Temporalio/Worker/ActivityWorker.cs (hunk 2 above)
Comment (cretz) [reply src/Temporalio/Worker/ActivityWorker.cs:679]: Fixed in next PR

File: src/Temporalio/Worker/WorkflowInstance.cs (hunk 12 above)
Comment (cretz) [reply src/Temporalio/Worker/WorkflowInstance.cs:658]: Correct
